Tests/CredentialsJWTTests/TestRawRouteJWT.swift

```diff
@@ -0,0 +1,289 @@
+            let profileInstance = try JWT<TestClaims>(jwtString: jwtString, verifier: .hs256(key: key!))
+            // An equivalent test profile, constructed directly.
+            let testTokenProfile = JWT(claims: TestClaims(sub: "Test"))
+            XCTAssertEqual(profileInstance.claims, testTokenProfile.claims, "The reference JWT instance did not match the instance decoded from the jwt string")
+        } catch {
+            XCTFail("error")
+        }
+    }
+
+    // Tests that when a correct token is supplied a valid UserProfile is created.
+    func testCorrectToken() {
+        performServerTest(router: router) { expectation in
+            self.performRequest(method: "get", path: "/rawtokenauth", callback: { response in
+                XCTAssertNotNil(response, "ERROR!!! ClientRequest response object was nil")
+                XCTAssertEqual(response?.statusCode, HTTPStatusCode.OK, "HTTP Status code was \(String(describing: response?.statusCode))")
+                do {
+                    guard let body = try response?.readString() else {
+                        XCTFail("No response body")
+                        return
+                    }
+                    let profile = body
+                    let testProfile = UserProfile(id: "Test", displayName: "Test", provider: "JWT")
+                    XCTAssertEqual(profile, testProfile.id)
+                } catch {
+                    XCTFail("Could not decode response: \(error)")
+                }
+                expectation.fulfill()
+            }, headers: ["X-Token-Type" : "JWT", "Authorization" : "Bearer " + self.jwtString])
+        }
+    }
+
+    // Tests that when an incorrect token is supplied a valid token but wrong UserProfile is created.
+    func testIncorrectToken() {
+        performServerTest(router: router) { expectation in
+            self.performRequest(method: "get", path: "/rawtokenauth", callback: { response in
+                XCTAssertNotNil(response, "ERROR!!! ClientRequest response object was nil")
+                XCTAssertEqual(response?.statusCode, HTTPStatusCode.OK, "HTTP Status code was \(String(describing: response?.statusCode))")
+                do {
+                    guard let body = try response?.readString() else {
+                        XCTFail("No response body")
+                        return
+                    }
+                    let profile = body
+                    let testProfile = UserProfile(id: "Test", displayName: "Test", provider: "JWT")
+                    XCTAssertNotEqual(profile, testProfile.id)
+                } catch {
+                    XCTFail("Could not decode response: \(error)")
+                }
+                expectation.fulfill()
+            }, headers: ["X-Token-Type" : "JWT", "Authorization" : "Bearer " + self.jwtString2])
+        }
+    }
+
+    // Tests that when an incorrect token is supplied an invalid token is supplied, user is unauthorized.
+    func testInvalidToken() {
```

Comment:
There are two types of invalid tokens: ones that don't verify (like this) and ones that are valid JWTs but don't contain the expected claims.  `testInvalidToken` tests the former.  I think there are two other tests we should have:

- `testWrongClaims()` where the token is a valid token, but the Claims are not `TestClaims`.  Perhaps something like `TestAlternateClaims` which has a claim `username` rather than `sub`.
- `testSubjectName()`, similar to `testCorrectToken()` which a JWT containing `TestAlternateClaims`, and where the subject option has been set to use `username` rather than `sub` (ie. test that customization facility actually works).